### PR TITLE
✨Allow `fake` ads in amp stories

### PR DESCRIPTION
**Pull request by calebcordry**
Before this change the ad would not render due to incorrect metadata around the `ctaType` and `ctaUrl`.

This allows a configuration like the following to render
```html
<amp-story-auto-ads>
  <script type="application/json">
    {
      "ad-attributes": {
        "type": "fake",
        "src": "/path/to/story/ad",
        "id": "i-amphtml-demo-fake",
        "a4a-conversion": true
      }
    }
  </script>
</amp-story-auto-ads>
```

### CODE REVIEW

**Comment by calebcordry** [comment]
cc/ @ithinkihaveacat @kul3r4

File: examples/amp-story/fake-ad.html [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+      poster-landscape-src="https://example.com/my-story/poster/4x3.jpg"
+      standalone>
+
+        <amp-story-auto-ads>
+            <script type="application/json">
+              {
+                "ad-attributes": {
+                  "type": "fake",
+                  "src": "/examples/amp-story/ads/app-install.html",
+                  "id": "i-amphtml-demo-fake",
```
Comment (zhouyx) [inline examples/amp-story/fake-ad.html]: Does amp-story-auto-ads check the validness of the id?

Comment (calebcordry) [reply examples/amp-story/fake-ad.html]: amp-story-auto ads just creates an `<amp-ad type="fake" id="i-amphtml..">`. There is no additional logic around checking the id, so it should behave the same as a normal "fake" ad. Are you worried about the doc being invalid?

Comment (calebcordry) [reply examples/amp-story/fake-ad.html]: After thinking about this more, I think I understand. The doc will be valid when it shouldn't be. I don't have a great solution. We could have the runtime not allow this in certain contexts, but I'm not sure what the signal would be. What do you think?

Comment (zhouyx) [reply examples/amp-story/fake-ad.html]: To allowing fake ad specifically we can have some kind of workaround. For example, one must put `id=i-amphtml-...` to the `<amp-story-auto-ads>` component. So the fake ad only loads for invalid AMP page. 

However i think this is related to #20003. Allowing developers to randomly assign attribute of the inserted ad in general sounds very unsafe to me. I can easily inject something like `i-amphtml-ghost` without validator catching that.

Comment (calebcordry) [reply examples/amp-story/fake-ad.html]: This is a good point. I agree a whitelist sounds good, but needs to be fleshed out. For instance `id` is allowed, but certain values are not. I will make some comments on the issue. 

To solve this problem I think that having an invalid signal on the amp-story-auto-ads tag is a good idea. I will refactor.

Comment (zhouyx) [reply examples/amp-story/fake-ad.html]: SG. Thank you

Comment (lannka) [reply examples/amp-story/fake-ad.html]: a different issue: are we inserting multiple `amp-ad` with the same ID?

**Comment by lannka** [comment]
@calebcordry what's the status of this PR?

**Comment by calebcordry** [comment]
I got distracted by other things :) I need to refactor this to make sure the doc is invalid when in use. This is very helpful for me / devrel when working on making these creatives so I will get it done.

File: examples/amp-story/fake-ad.html (hunk 1 above)
Comment (calebcordry) [reply examples/amp-story/fake-ad.html]: yes, is this not ok for testing?

Comment (calebcordry) [reply examples/amp-story/fake-ad.html]: changed this around a bit so auto ads injects a unique id. Pubs just add the attribute to the auto-ads tag itself.

**Comment by calebcordry** [comment]
@zhouyx Could you re-review when you have a chance?

File: examples/amp-story/ads/app-install.html [hunk 2]
```diff
@@ -0,0 +1,101 @@
+<!doctype html>
```
Comment (torch2424) [inline examples/amp-story/ads/app-install.html]: How is this example related to the fake ads? 🤔 No worries if it isn't just curious 😄

File: extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js [hunk 3]
```diff
@@ -334,6 +336,10 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
    */
   schedulePage_() {
     const page = this.createAdPage_();
+    if (!page) {
+      // Creation of ad has failed.
```
Comment (torch2424) [inline extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: Should we throw an error or warning here? 🤔 

It may be a bit noisy though if we did.

Comment (calebcordry) [reply extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: I think this should only happen in the fake case which will already user().warn()

File: examples/amp-story/ads/app-install.html (hunk 2 above)
Comment (calebcordry) [reply examples/amp-story/ads/app-install.html]: This is the fake ad that gets loaded.

File: extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js (hunk 3 above)
Comment (torch2424) [reply extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: Sounds good to me! 😄

File: examples/amp-story/ads/app-install.html (hunk 2 above)
Comment (torch2424) [reply examples/amp-story/ads/app-install.html]: Oh haha! Rad! Sounds good 😄

File: extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js [hunk 4]
```diff
@@ -449,8 +455,18 @@ export class AmpStoryAutoAds extends AMP.BaseElement {
...
-      `"${configAttrs.type}" ad type is not supported`);
+    const {type} = configAttrs;
+    userAssert(!!ALLOWED_AD_TYPES[type], `${TAG}: ` +
+      `"${type}" ad type is not supported`);
+
+    if (type === 'fake') {
+      const id = this.element.getAttribute('id');
+      if (!id || !startsWith(id, 'i-amphtml-demo-')) {
+        user().warn(TAG, 'id must start with i-amphtml-demo- to use fake ads');
+        return null;
```
Comment (zhouyx) [inline extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: Have we handle the case where `this.createAdElement_` returns null?
https://github.com/ampproject/amphtml/blob/391a431816894cc726e8c7891081885589f391bf/extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js#L356

Comment (calebcordry) [reply extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: Nope you are right, null check was in wrong place. Instead of adding a bunch of logic for fake ad destruction, what do you think about creating a dummy element instead like `<p>Fake ad discarded.<p>` and inserting that?

Comment (zhouyx) [reply extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: What about we keep it simple, and use `userAssert()` instead?

Comment (calebcordry) [reply extensions/amp-story-auto-ads/0.1/amp-story-auto-ads.js]: Good idea. Done.

**Comment by zhouyx** [comment]
LGTM
Depending on who you expect to use this feature, it may worth documenting (doc or error message) that we require an invalid AMP doc to use fake ad.
